src/server/qgsserviceregistry.cpp

```diff
@@ -0,0 +1,185 @@
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgsserviceregistry.h"
+#include "qgsservice.h"
+#include "qgslogger.h"
+#include "qgsmessagelog.h"
+
+namespace {
+
+    // Build a key entry from name and version
+    QString make_service_key( const QString& name, const QString& version  )
+    {
+        return QString( "%1_%2" ).arg(name,version);
+    }
+
+    // Compare two version strings:
+    // The strings are splitted into dot separated segment
+    // Each segment are compared up to the shortest number of segment of the 
+    // lists. Remaining segments are dropped.
+    // If both segments can be intepreted as numbers the are compared as numbers, otherwise 
+    // They are compared lexicographically.
+    // Return true if v1 is greater than v2
+    bool is_version_greater( const QString& v1, const QString& v2 )
+    {
+        QStringList l1 = v1.split('.');
+        QStringList l2 = v2.split('.');
+        QStringList::iterator it1 = l1.begin();
+        QStringList::iterator it2 = l2.begin();
+        bool isint;
+        while( it1 != l1.end() && it2 != l2.end() ) 
+        {
+            if ( *it1 != *it2 )
+            {
+                // Compare as numbers
+                int i1 = it1->toInt(&isint);
+                if(isint) 
+                {
+                    int i2 = it2->toInt(&isint);
+                    if( isint && i1 != i2 )
+                    {
+                        return i1 > i2;
+                    }
+                }
+                // Compare lexicographically
+                if ( !isint )
+                {
+                    return *it1 > *it2;
+                }
+            }
+            ++it1;
+            ++it2;
+        }
+        // We reach the end of one of the list 
+        return false;
+    }
+    
+} // namespace
+
+class QgsServiceEntry
+{
+    public:
+       ~QgsServiceEntry() 
+        {
+            // We have the ownership by design
+            // XXX Take care of /Transfer/ decorator for registerService in sip 
+            QgsDebugMsg( QString("Deleting service %1 %2").arg(mName, mVersion) );
+            delete mService;
+        }
+
+       QgsServiceEntry( const QString& name, QgsService* service, const QString& version )
+       : mName(name)
+       , mService(service)
+       , mVersion(version)
+       {}
+
+       QString     mName;
+       QgsService* mService;
+       QString     mVersion;    
+};
```

Comment:
Maybe these bits of information like name/version could be already inside of QgsService class, so there is no need for service entry classes?